// @assistant-ui/react-ai-sdk/dist/ui/use-chat/AssistantChatTransport.js
import { DefaultChatTransport, } from "ai";
import { toToolsJSONSchema } from "assistant-stream";
export class AssistantChatTransport extends DefaultChatTransport {
    runtime;
    constructor(initOptions) {
        super({
            ...initOptions,
            prepareSendMessagesRequest: async (options) => {
                const context = this.runtime?.thread.getModelContext();
                const id = (await this.runtime?.threads.mainItem.initialize())?.remoteId ??
                    options.id;
                const optionsEx = {
                    ...options,
                    body: {
                        callSettings: context?.callSettings,
                        system: context?.system,
                        config: context?.config,
                        tools: toToolsJSONSchema(context?.tools ?? {}),
                        ...options?.body,
                    },
                };
                const preparedRequest = await initOptions?.prepareSendMessagesRequest?.(optionsEx);
                return {
                    ...preparedRequest,
                    body: preparedRequest?.body ?? {
                        ...optionsEx.body,
                        id,
                        messages: options.messages,
                        trigger: options.trigger,
                        messageId: options.messageId,
                        metadata: options.requestMetadata,
                    },
                };
            },
        });
    }
    setRuntime(runtime) {
        this.runtime = runtime;
    }
}
