// @assistant-ui/react/dist/index.js
import "./types/store-augmentation.js";
// Re-export from @assistant-ui/store
export { useAui, AuiProvider, useAuiState, useAuiEvent, AuiIf, } from "@assistant-ui/store";
export * from "./legacy-runtime/runtime/index.js";
export * from "./legacy-runtime/cloud/index.js";
export * from "./legacy-runtime/runtime-cores/index.js";
export * from "./context/index.js";
export * from "./model-context/index.js";
export * from "./primitives/index.js";
export * from "./types/index.js";
export * from "./devtools/index.js";
export * from "./client/index.js";
export * as INTERNAL from "./internal.js";
// ============================================================================
// Backwards compatibility - deprecated exports
// ============================================================================
/**
 * @deprecated Use `useAui` instead. This alias will be removed in v0.13.
 */
export { useAui as useAssistantApi } from "@assistant-ui/store";
/**
 * @deprecated Use `useAuiState` instead. This alias will be removed in v0.13.
 */
export { useAuiState as useAssistantState } from "@assistant-ui/store";
/**
 * @deprecated Use `useAuiEvent` instead. This alias will be removed in v0.13.
 */
export { useAuiEvent as useAssistantEvent } from "@assistant-ui/store";
/**
 * @deprecated Use `AuiIf` instead. This alias will be removed in v0.13.
 */
export { AuiIf as AssistantIf } from "@assistant-ui/store";
